third-party extension support for typedoc plugin

boneskull opened the pull request:
WIP PR which makes changes needed for `@appium/typedoc-plugin-appium` to work with "third party" extensions (in other words: extensions not in this monorepo).

## Summary

E2E tests are quasi-blocking on merge & release of https://github.com/TypeStrong/typedoc/issues/2130; it is a huge pain to test without this.

- Adds some options to control output
- Rename `CommandsReflection` to `ExtensionReflection` due to poor granular control over the display name
- Rename the custom "kinds"; add kinds for drivers and plugins (which affects display titles)
- Remove namespace from "kinds" for same reason
- Refactors, reformatting
- Rename entry point from `plugin.ts` to `index.ts`
- Update keywords in `package.json` for auto-discovery
- Remove cruft from root `typedoc.json`
- Update `README.md`

boneskull committed:
fix(typedoc-plugin-appium): bad directory in repository url

boneskull committed:
fix(typedoc-plugin-appium): add peer dep of appium

boneskull committed:
chore(tsconfig): strip internal stuff from declarations

typedoc respects this flag as well

boneskull committed:
feat(typedoc-plugin-appium): add options and 3p support

E2E tests are quasi-blocking on merge & release of https://github.com/TypeStrong/typedoc/issues/2130; it is a huge pain to test without this.

- Adds some options to control output
- Rename `CommandsReflection` to `ExtensionReflection` due to poor granular control over the display name
- Rename the custom "kinds"; add kinds for drivers and plugins (which affects display titles)
- Remove namespace from "kinds" for same reason
- Refactors, reformatting
- Rename entry point from `plugin.ts` to `index.ts`
- Update keywords in `package.json` for auto-discovery
- Remove cruft from root `typedoc.json`
- Update `README.md`
- Update peer dependencies; the other plugins must be peer deps or typedoc will be unable to auto-discover them

boneskull commented:
Current dependencies on/for this PR:
* master
  * **PR #17971** <a href="https://app.graphite.dev/github/pr/appium/appium/17971" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #17992** <a href="https://app.graphite.dev/github/pr/appium/appium/17992" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 
      * **PR #18007** <a href="https://app.graphite.dev/github/pr/appium/appium/18007" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a>  👈
    * **PR #17986** <a href="https://app.graphite.dev/github/pr/appium/appium/17986" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/appium/appium/18007?utm_source=stack-comment).

jlipps reviewed (approved):
seems reasonable to me!

boneskull committed:
fix(typedoc-plugin-appium): allow commands as properties referencing async methods

Instead of only inspecting `ReflectionKind.Method`, we now examine `ReflectionKind.Property`, which can be of type `ReflectionType` pointing to an async function.  Example:

```js
async function baz() {}

class Foo {
  async bar() {} // ReflectionKind.Method
  baz = baz; // ReflectionKind.Property referencing a function
}
```

Also ensures `static` methods aren't picked up; this could happen if a static method shares the same name with a known command method as defined in the builtin routes.

boneskull committed:
chore(typedoc-plugin-appium): tweaks

- `BuiltinMethodMapConverter`'s `convert()` will _always_ return a `BuiltinCommands` / removed unused import
- Verbose logging for skipped methods
- Remove useless type aliases from option declarations and add some no-op typechecks (these will be compiled away)
- Add some docstrings to guards and types
- `CallSignatureReflectionWithParams` is now `CallSignatureReflectionWithArity`, and it is specific about having a nonzero arity. Broke out `CallSignatureReflection` from this, since the guard and type didn't fully match up

boneskull commented:
will merge once CI passes
